content/blog/withr-3-0-0/index.Rmd

```diff
@@ -0,0 +1,228 @@
+  author: Neal E. Johnson
+
+categories: [package]
+tags: [r-lib, withr]
+---
+
+It's not without jubilant bearing that we announce the release of the 3.0.0 version of [withr](https://withr.r-lib.org/), the tidyverse solution for automatic cleanup of resources! In this release, the internals of withr were rewritten to improve the performance and increase the compatibility with base R's `on.exit()` mechanism.
+
+You can install it from CRAN with:
+
+```{r, eval = FALSE}
+install.packages("withr")
+```
+
+In this blog post we'll go over the changes that made this rewrite possible, but first we'll rewiew the cleanup strategies made possible by withr.
+
+You can see a full list of changes in the [release notes](https://withr.r-lib.org/news/index.html#withr-300).
+
+```{r setup, echo = FALSE}
+# To avoid distracting calls to `eval()` in `local()`
+stop <- rlang::abort
+```
+
+
+## Cleaning up resources with base R and with withr
+
+Traditionally, resource cleanups in R is done with `base::on.exit()`. Cleaning up in the on-exit hook ensures that the cleanup happens both in the normal case, when the code has finished running without error, and in the error case, when something went wrong and execution is interrupted.
+
+`on.exit()` is meant to be used inside functions but it also works within `local()`, which we'll use here for our examples:
+
+```{r}
+local({
+  on.exit(message("Cleaning time!"))
+  1 + 2
+})
+```
+
+```{r, error = TRUE}
+local({
+  on.exit(message("Cleaning time!"))
+  stop("uh oh")
+  1 + 2
+})
+```
+
+`on.exit()` is guaranteed to run no matter what and this property makes it invaluable for resource cleaning. No more accidental littering!
+
+However the process of cleaning up this way can be a bit verbose and feel too manual. Here is how you'd create and clean up a temporary file for instance:
+
+```{r}
+local({
+  my_file <- tempfile()
+
+  file.create(my_file)
+  on.exit(file.remove(my_file))
+
+  writeLines(c("a", "b"), con = my_file)
+})
+```
+
+To streamline the process, withr provides `local_`-prefixed tools that combine both the creation or modification of a resource and its restoration to the original state in a single function:
+
+```{r}
+local({
+  my_file <- withr::local_tempfile()
+
+  writeLines(c("a", "b"), con = my_file)
+})
+```
+
+In this case we have created a resource (a file), but the same principle applies to modifying resources such as global options:
+
+```{r}
+local({
+  # Let's temporarily print with a single decimal place
+  withr::local_options(digits = 1)
+  print(1/3)
+})
+
+# The original option value has been restored
+getOption("digits")
+
+print(1/3)
+```
+
+And you can equivalently use the `with_`-prefixed variants (from which the package takes his name!), this way you don't need to wrap in `local()`:
+
+```{r}
+withr::with_options(list(digits = 1), print(1/3))
+```
+
+The `with_` functions are useful for creating very small scopes for given resources, inside or outside a function.
```

Comment:
I'm not 100% sure it's even worth mentioning `with_` here given how rarely it's used.

Reply:
I added that sentence to explain why withr is called withr. But I guess we can keep it mysterious.